src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentCodeActionProvider.cs

```diff
@@ -51,31 +56,51 @@ public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeAct
             return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
         }
 
-        var (startElementNode, endElementNode) = GetStartAndEndElements(context, syntaxTree, _logger);
-
         // Make sure the selection starts on an element tag
+        var (startElementNode, endElementNode) = GetStartAndEndElements(context, syntaxTree, _logger);
         if (startElementNode is null)
         {
             return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
         }
 
+        if (endElementNode is null)
+        {
+            endElementNode = startElementNode;
+        }
+
         if (!TryGetNamespace(context.CodeDocument, out var @namespace))
         {
             return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
         }
 
-        var actionParams = CreateInitialActionParams(context, startElementNode, @namespace);
+        var actionParams = CreateInitialActionParams(context, startElementNode, @namespace);        
 
         ProcessSelection(startElementNode, endElementNode, actionParams);
 
+        var utilityScanRoot = FindNearestCommonAncestor(startElementNode, endElementNode) ?? startElementNode;
+
+        // The new component usings are going to be a subset of the usings in the source razor file
+        var usingsInFile = syntaxTree.Root.DescendantNodes()
+            .OfType<CSharpStatementLiteralSyntax>()
+            .Where(literal => literal.SerializedValue.Contains("using"))
+            .Select(literal => literal.SerializedValue)
```

Comment:
Close but this is still wrong because this will be true for statements where a variable has `using` in the name (and worth adding a test for that case). I _think_ this is correct but I'm coding in github so 🤷

```suggestion
            .OfType<CSharpStatementLiteralSyntax>()
            .Where(statement => statement.LiteralTokens.FirstOrDefault() is SyntaxToken { Kind: SyntaxKind.Keyword } token && CSharpTokenizer.GetTokenKeyword(token) == CSharpSyntaxKind.UsingKeyword))
```

Reply:
Ahhh, you're right